Management Command tool to scrub references to the username in form metadata/xml

pr33thi opened the pull request:
Task #2 in this doc:
https://docs.google.com/document/d/1qCtWmQ2disUEi86Ty5JwNS_3rFiveMDpZf0cj0zVcvQ/edit?ts=5aafed1b

- Gets rid of the username from the xml (and therefore the metadata) (SQL)
- Gets rid of the username from the form metadata (couch)
^^Replace username in both cases with "Redacted User (GDPR)"

Note: Saving the operation doesn't work/ doesn't add a form history tab, feedback on that would be great @snopoke or anyone familiar

pr33thi committed:
Make management command to iterate through all forms and update their username with a dummy username

pr33thi committed:
Refactor code to separate out both replacements

pr33thi committed:
Update username in xml and in json metadata

pr33thi committed:
Delete old database entry if write_content is actually overwriting something (untested)

pr33thi committed:
Delete old database entry when calling wrote_content on an attachment

pr33thi committed:
Add comments

pr33thi committed:
Delete debug statement from write_content

pr33thi committed:
Pass in username and domain as command line args

pr33thi committed:
Modify code for tests

pr33thi committed:
Update xml namespace to 0

pr33thi committed:
Fix updated xml, finish xml unit test

pr33thi committed:
Test updated xml in database

pr33thi committed:
More couch test, updated xml and its now valid

pr33thi committed:
Test replacing metadata for couch and SQL

pr33thi committed:
Refactor common code out of test cases

pr33thi committed:
Clean up tests

pr33thi committed:
Merge branch 'master' into pv/un-scrub-gdpr-272917

pr33thi committed:
Record operation in form history

pr33thi committed:
Refactored xml modification into separate dbaccessor classe

pr33thi committed:
Refactor to remove repeated test logic

pr33thi committed:
Use a custom simple form for GDPR tests

pr33thi committed:
Fix pep8 issues

pr33thi committed:
Merge branch 'master' into pv/un-scrub-gdpr-272917

pr33thi committed:
Remove extra db call causing unit tests to fail

pr33thi committed:
Remove extra db call causing unit tests to fail

pr33thi committed:
Merge branch 'pv/un-scrub-gdpr-272917' of https://github.com/dimagi/commcare-hq into pv/un-scrub-gdpr-272917

pr33thi committed:
cleanup

pr33thi committed:
Reformat for PEP8

pr33thi committed:
Merge branch 'pv/un-scrub-gdpr-272917' of https://github.com/dimagi/commcare-hq into pv/un-scrub-gdpr-272917

millerdev commented on corehq/apps/users/management/commands/gdpr_scrub_user.py:
```diff
@@ -0,0 +1,36 @@
+        # Replace the old username with the new username
+        form_attachment_dict["data"]["n0:meta"]["n0:username"] = new_username
+        # Convert the dict back to xml
+        form_attachment_xml_new = xmltodict.unparse(form_attachment_dict)
```
How similar is the new XML to the original? Are all namespaces and other XML features preserved? This seems like a rather complex and possibly lossy operation, and I'm surprised it can be done so simply given that XML does not really have a canonical form. I am not familiar with `xmltodict` at all, fwiw.

pr33thi replied:
Switched form xmltodict to lxml here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/c7b17ed7e4ae91458fd2fdbad435222bceeb921e

millerdev commented on corehq/apps/users/management/commands/gdpr_scrub_user.py:
```diff
@@ -0,0 +1,36 @@
+            this_form_accessor.modify_attachment_xml_and_metadata(form_data, form_attachment_xml_new)
+
+    @staticmethod
+    def parse_form_data(form_data, new_username):
```
Nitpick: what do you think of renaming this to `update_form_data` since it doesn't just parse the form?

pr33thi replied:
Good idea, changed here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/8d6837e0473a99499db893b8e79ac0d7e1afb62b

millerdev commented on corehq/form_processor/tests/test_gdpr_scrub_user.py:
```diff
@@ -0,0 +1,88 @@
+        FormAccessors(DOMAIN).modify_attachment_xml_and_metadata(form, new_form_xml)
+
+        # Test that the xml changed
+        form_attachment_xml = form.get_attachment("form.xml")
```
Would be nice to compare this new/updated XML to the original `simple_form` (mutated with the new username) to ensure that the new XML matches the old using something different than the library that you're making the original mutation with. We have [an assert method that does that](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/tests/util.py#L30).

pr33thi replied:
Fixed here (was much easier with lxml instead of xmltodict):
https://github.com/dimagi/commcare-hq/pull/19979/commits/6e77f07d8078d49dc91fdfed0c281488ab6a508e

millerdev reviewed (commented):
Looking good. I added a few comments.

pr33thi committed:
Change gdpr_scrub_user to gdpr_scrub_user_from_form

pr33thi committed:
Replace xmltodict with etree

pr33thi committed:
Replace xmltodict with etree

pr33thi committed:
Add fake_instance_id to xml

pr33thi committed:
Add fake_instance_id to xml

pr33thi committed:
Relaced xmltodict

pr33thi committed:
Merge branch 'master' into pv/un-scrub-gdpr-272917

pr33thi committed:
Replace xmltodict with lxml in the code and the unit tests

pr33thi commented on corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py:
```diff
@@ -0,0 +1,36 @@
+from __future__ import unicode_literals
+from django.core.management.base import BaseCommand
+from corehq.form_processor.interfaces.dbaccessors import FormAccessors
+from StringIO import StringIO
```
@nickpell 
what does this mean?

nickpell replied:
You'll want to replace ```from StringIO import StringIO``` with ```from io import StringIO``` (or ```BytesIO``` if you're dealing with byte strings).

Python 3 uses the ```io``` module, and removes the ```StringIO``` module.

pr33thi committed:
Take in username instead of iser_id, improve error handling and logging

pr33thi committed:
Formatting and py3 compatability

pr33thi committed:
Change test name to match file being tested

pr33thi committed:
Convert xml to unicode before running lxml commands on it

pr33thi committed:
Remove print statement

pr33thi committed:
Uncomment test

pr33thi committed:
Replace unicode with six.text_types

millerdev commented on corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py:
```diff
@@ -0,0 +1,50 @@
+        user = CommCareUser.get_by_username(username)
+        if not user:
+            logger.info("User {} not found.".format(username))
+            sys.exit()
```
should say `sys.exit(1)`, which indicates an error condition

pr33thi replied:
https://github.com/dimagi/commcare-hq/pull/19979/commits/4b45c9117c40d43ca06f3d5dfeaae87f0d68cc38

millerdev commented on corehq/form_processor/backends/couch/dbaccessors.py:
```diff
@@ -126,6 +131,14 @@ def _form_undelete(doc):
+        form_data.put_attachment(form_attachment_new_xml, name="form.xml", content_type='text/xml')
+        operation = XFormOperation(user_id=SYSTEM_USER_ID, date=datetime.utcnow(),
+                                   operation='Scrub username for GDPR compliance.')
+        form_data.history.append(operation)  # TODO: should this show in Form History tab? it doesn't
```
I think this TODO should be resolved before the PR is merged if possible. What do you think?

proteusvacuum replied:
check out this comment to see how you might achieve this: https://github.com/dimagi/commcare-hq/pull/19973#discussion_r177989201

snopoke replied:
This is operating on the couch models so this approach should be fine. A unit test should be easy to make sure it's working

pr33thi replied:
Tested here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/8f4cf33e4b0319561c0110c5bb470e9790b3bb27

millerdev commented on corehq/form_processor/tests/test_gdpr_scrub_user_from_forms.py:
```diff
@@ -0,0 +1,93 @@
+        actual_form_xml_from_db = XFormAttachmentSQL.read_content(attachment_metadata)
+        self.assertXMLEqual(EXPECTED_FORM_XML, actual_form_xml_from_db)
+
+        # TODO: Test that the operations history is updated
```
Are you planning to do this before the PR is merged?

pr33thi replied:
Yep! I mentioned I'm going to finish this in the PR description (but that's way up there now)

millerdev reviewed (commented):
One nitpicky comment and two questions. Looking good.

esoergel commented on corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py:
```diff
@@ -0,0 +1,50 @@
+            form_attachment_xml_new = self.update_form_data(form_data, new_username)
+            this_form_accessor.modify_attachment_xml_and_metadata(form_data, form_attachment_xml_new)
+            forms_updated += 1
+        logging.info("Updated {} forms for user {}".format(forms_updated, username))
```
What do you think about putting in a confirmation here?  Something like "Update 132 forms on 'domain' for username@domain.commcarehq.org? (Y/n)".  These commands are intentionally destructive and misuses would be unrecoverable, so we should be pretty cautious.

pr33thi replied:
Good call, added it in here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/dd1468d1e5d288001f6157d643795b415839b499

esoergel commented on corehq/form_processor/backends/couch/dbaccessors.py:
```diff
@@ -102,6 +103,10 @@ def get_attachment_content(form_id, attachment_id):
 
+    @staticmethod
+    def update_form(form):
+        form.save()
```
Is this (and the superclass version) used anywhere?

pr33thi replied:
Nope! This is a remnant from something I tried earlier that isn't relevant anymore. 
Removed them both here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/3785fe4ab63bca7fa19dd033723a144151077093

esoergel commented on corehq/form_processor/tests/test_gdpr_scrub_user_from_forms.py:
```diff
@@ -0,0 +1,93 @@
+        <n1:appVersion xmlns:n1="http://commcarehq.org/xforms"></n1:appVersion>
+    </n0:meta>
+    {case_block}
+</data>"""
```
Any reason not to just use the existing `SIMPLE_FORM` template?  Looks like the differences are `n0` instead of `n1`, and the `instanceID`, which both seems tangential to this test.

esoergel reviewed (approved):
Is there going to be a second command to remove the other references to `username` on the form db object?

pr33thi committed:
Change sys.exit() to sys.exit(1)

pr33thi committed:
Add confirmation message before changing username in forms

pr33thi committed:
Change form to form(s) in logging statement

snopoke commented on corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py:
```diff
@@ -0,0 +1,50 @@
+        logging.info("Updated {} forms for user {}".format(forms_updated, username))
+
+    @staticmethod
+    def update_form_data(form_data, new_username):
```
a string search and replace might be faster

pr33thi replied:
True, I was imagining a dev would just run this in the background while doing other things anyway (as it'll probably end up taking a while either way) and searching the xml seems a little more robust.

snopoke commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -581,6 +582,17 @@ def soft_undelete_forms(domain, form_ids):
+    def modify_attachment_xml_and_metadata(form_data, form_attachment_new_xml):
+        attachment_metadata = form_data.get_attachment_meta("form.xml")
+        # Write the new xml to the database
+        XFormAttachmentSQL.write_content(attachment_metadata, form_attachment_new_xml)
```
why not `attachment_metadata.write_content(form_attachment_new_xml)`

pr33thi replied:
True, fixed here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/d42272370e58270d60fc2b82815e0ad48fdab195

snopoke commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -581,6 +582,17 @@ def soft_undelete_forms(domain, form_ids):
+        attachment_metadata.save()
+        operation = XFormOperationSQL(user_id=SYSTEM_USER_ID, date=datetime.utcnow(),
+                                      operation='Scrub username for GDPR compliance.')
+        form_data.history.append(operation)  # TODO: should this show in Form History tab? it doesn't
```
this won't work for SQL. Rather do this:
```python
form_data.track_create(operation)
FormAccessorSQL.update_form(form_data)
```

pr33thi replied:
Updated here (and did some extra test stuff which I clean up in the below pr):
https://github.com/dimagi/commcare-hq/pull/19979/commits/c38433c08998c958cf9e44cdd6f5a26248314aa8

And tested here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/8f4cf33e4b0319561c0110c5bb470e9790b3bb27

pr33thi committed:
Remove update_form from couch accessor and superclass

pr33thi committed:
Using attachment_metadata.write_content(form_attachment_new_xml)

nickpell commented on corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py:
```diff
@@ -0,0 +1,54 @@
+        user_id = user._id
+        form_ids = this_form_accessor.get_form_ids_for_user(user_id)
+        new_username = "Redacted User (GDPR)"
+        input_response = raw_input("Update {} form(s) for user {} in domain {}? (y/n): ".format(len(form_ids), username, domain))
```
This is fixed with ```six```: https://pythonhosted.org/six/#module-six.moves

pr33thi replied:
Thanks! Fixed here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/77a4504e821f0beef78f42976bc50fb547df0afb

pr33thi committed:
Separate out tests into 3 separate classes

pr33thi committed:
Merge branch 'master' into pv/un-scrub-gdpr-272917

pr33thi committed:
Fix Operations code, and add to tests

pr33thi committed:
Remove extra lines from output format and unneeded encoding

snopoke commented on corehq/form_processor/models.py:
```diff
@@ -343,7 +343,7 @@ def form_data(self):
         """:returns: List of XFormOperationSQL objects"""
         from corehq.form_processor.backends.sql.dbaccessors import FormAccessorSQL
-        operations = FormAccessorSQL.get_form_operations(self.form_id) if self.is_saved() else []
+        operations = FormAccessorSQL.get_form_operations(self.form_id) if self.is_saved() else ['hi!!!!']
```
`hi!!!!`

pr33thi replied:
oops, fixed here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/94ca91c82202a27348915facb2c53e01b4c1c263

snopoke commented on corehq/form_processor/tests/test_gdpr_scrub_user_from_forms.py:
```diff
@@ -0,0 +1,118 @@
+        self.db.close()
+        super(GDPRScrubUserFromFormsCouchTests, self).tearDown()
+
+    def test_modify_attachment_xml_and_metadata_couch(self):
```
These tests look the same. The common pattern is to do this:

```
class CouchTests(TestCase):
    def my_test(self):
        ....

@use_sql_backend
class SQLtests(CouchTests)
    pass


```

pr33thi replied:
cool, didn't know I could do that, fixed here:

https://github.com/dimagi/commcare-hq/pull/19979/commits/34a94c2a68078a5d77e1bbbaad68465d5120f2d9

millerdev reviewed (commented):
👍 after @snopoke's feedback is addressed

pr33thi committed:
Replace raw_input with py3 compatible six call

pr33thi committed:
Stickler

pr33thi committed:
Create an operation type within XFormOperationSQL

pr33thi committed:
Remove debug text from history property

pr33thi committed:
Run similar test with subclassing

pr33thi commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -581,6 +582,17 @@ def soft_undelete_forms(domain, form_ids):
+        attachment_metadata.write_content(form_attachment_new_xml)
+        attachment_metadata.save()
+        operation = XFormOperationSQL(user_id=SYSTEM_USER_ID, date=datetime.utcnow(),
+                                      operation='Scrub username for GDPR compliance.')
```
Ah, didnt know about that, using that logic here:

https://github.com/dimagi/commcare-hq/pull/19979/commits/8da6f23e2660f1d6ed3e0c1cd86a6965055c3f43

And just to be clear, `XFormOperation` doesn't have those terms predefined, right? I just passed in `'gdpr_scrub'` directly here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/34a94c2a68078a5d77e1bbbaad68465d5120f2d9#diff-ff25fb2af7d300989db9379ec9dd37fcL134

pr33thi committed:
Stickler

snopoke commented on corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py:
```diff
@@ -0,0 +1,55 @@
+
+    def handle(self, username, domain, **options):
+        this_form_accessor = FormAccessors(domain=domain)
+        user = CommCareUser.get_by_username(username)
```
This could be a web user too. Probably better to use `CouchUser.get_by_username`

pr33thi replied:
Fixed here (and learned the differences between types of users +1)
https://github.com/dimagi/commcare-hq/pull/19979/commits/de2e98ad22ae74f559608bbb2f0a2194a7599cdf

snopoke commented on corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py:
```diff
@@ -0,0 +1,55 @@
+            sys.exit(1)
+        user_id = user._id
+        form_ids = this_form_accessor.get_form_ids_for_user(user_id)
+        new_username = "Redacted User (GDPR)"
```
might be nice to have this as a constant

pr33thi replied:
Done here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/9782b6ada66f767615c35af61d31483a9a562b70

snopoke commented on corehq/apps/users/management/commands/gdpr_scrub_user_from_forms.py:
```diff
@@ -0,0 +1,55 @@
+        xml_elem = etree.parse(StringIO(six.text_type(form_attachment_xml)))
+        id_elem = xml_elem.find("{http://openrosa.org/jr/xforms}meta").find(
+            "{http://openrosa.org/jr/xforms}username")
+        id_elem.text = new_username
```
I still think string search and replace would be faster and more robust here. It is possible that a username could be saved to any question in a form.

millerdev replied:
Weighing in here because I think an unstructured search/replace has some significant disadvantages. Suppose a project is scrubbing username `test@test.com` and there's another user in the domain named `my_test@test.com`—unless crafted very carefully (and email addresses are notoriously hard to parse) a search/replace may match things that should not be replaced. Also, unstructured search/replace could cause other parts of the metadata (or anything stored in couch or SQL or elastic that has been parsed from the XML) to become out of sync with the XML.

pr33thi replied:
Agreed ^, besides I checked with @krsdimagi who confirmed that instances of the username not specifically in the `username` tag are not in our scope

snopoke commented on corehq/form_processor/backends/couch/dbaccessors.py:
```diff
@@ -126,6 +127,14 @@ def _form_undelete(doc):
+        form_data.put_attachment(form_attachment_new_xml, name="form.xml", content_type='text/xml')
+        operation = XFormOperation(user_id=SYSTEM_USER_ID, date=datetime.utcnow(),
+                                   operation='gdpr_scrub')
+        form_data.history.append(operation)  # TODO: should this show in Form History tab? it doesn't
```
is this comment still true?

pr33thi replied:
Nope, thanks for catching that
Removed here:
https://github.com/dimagi/commcare-hq/pull/19979/commits/39339c57c1a196f76b02b81d5f29ed0e4859a9b7

snopoke commented on corehq/form_processor/backends/couch/dbaccessors.py:
```diff
@@ -126,6 +127,14 @@ def _form_undelete(doc):
         return _soft_undelete(XFormInstance.get_db(), form_ids, _form_undelete)
 
+    @staticmethod
+    def modify_attachment_xml_and_metadata(form_data, form_attachment_new_xml):
```
this still needs to modify the form metadata for couch forms

pr33thi replied:
Fixed across these 2 commits:

https://github.com/dimagi/commcare-hq/pull/19979/commits/39575949e103611d51e4b0a962ffb7528dda8467

https://github.com/dimagi/commcare-hq/pull/19979/commits/856c22edd5a2e6a8def96fa90f4510f04da6754d

pr33thi committed:
Merge branch 'master' into pv/un-scrub-gdpr-272917

pr33thi committed:
Delete outdated TODO

pr33thi committed:
Make NEW_USERNAME constant

pr33thi committed:
Change CommCareUser to CouchUser

pr33thi committed:
Using regex to replace all username instances instead of lxml forjust username tag

Using regex to replace all username instances instead of lxml forjust username tag

pr33thi committed:
Add test for changed username in metadata

pr33thi committed:
Attempt to update couch metadata

pr33thi committed:
Update metadata in couch

pr33thi committed:
Use lxml instead of regex for username replacement

pr33thi committed:
Update tests

pr33thi committed:
Finish resetting back to lxml

pr33thi committed:
Cleanup code

pr33thi committed:
Stickler

millerdev commented on corehq/form_processor/backends/couch/dbaccessors.py:
```diff
@@ -126,6 +127,18 @@ def _form_undelete(doc):
+        operation = XFormOperation(user_id=SYSTEM_USER_ID, date=datetime.utcnow(),
+                                   operation='gdpr_scrub')
+        form_data.history.append(operation)
+        form_data.save()
```
nit: this save could be omitted since it's saved again just below.

snopoke reviewed (approved):
pending resolution to double save comment: https://github.com/dimagi/commcare-hq/pull/19979/files#r179637693

pr33thi committed:
Delete extra form_data.save()

millerdev commented:
🎉
